providers/implementations/ciphers/cipher_chacha20_poly1305.c.in

```diff
@@ -103,181 +100,164 @@ static int chacha20_poly1305_get_params(OSSL_PARAM params[])
                                           CHACHA20_POLY1305_IVLEN * 8);
 }
 
+{- produce_param_decoder('chacha20_poly1305_get_ctx_params',
+                         (['CIPHER_PARAM_KEYLEN',            'keylen', 'size_t'],
+                          ['CIPHER_PARAM_IVLEN',             'ivlen',  'size_t'],
+                          ['CIPHER_PARAM_AEAD_TAGLEN',       'taglen', 'size_t'],
+                          ['CIPHER_PARAM_AEAD_TAG',          'tag',    'octet_string'],
+                          ['CIPHER_PARAM_AEAD_TLS1_AAD_PAD', 'pad',    'size_t'],
+                         )); -}
+
 static int chacha20_poly1305_get_ctx_params(void *vctx, OSSL_PARAM params[])
 {
     PROV_CHACHA20_POLY1305_CTX *ctx = (PROV_CHACHA20_POLY1305_CTX *)vctx;
-    OSSL_PARAM *p;
+    struct chacha20_poly1305_get_ctx_params_st p;
 
-    for (p = params; p->key != NULL; p++)
-        switch (chacha20_poly1305_get_ctx_params_find_pidx(p->key)) {
-        default:
-            break;
+    p = chacha20_poly1305_get_ctx_params_decoder(params);
 
-        case PIDX_CIPHER_PARAM_IVLEN:
-            if (!OSSL_PARAM_set_size_t(p, CHACHA20_POLY1305_IVLEN)) {
-                ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
-                return 0;
-            }
-            break;
+    if (p.ivlen != NULL
```

Comment:
I wonder if these are such common cases that they will be macroed at some point.